cql3/statements/create_table_statement.cc

```diff
@@ -193,6 +193,13 @@ std::unique_ptr<prepared_statement> create_table_statement::raw_statement::prepa
 
     auto stmt = ::make_shared<create_table_statement>(*_cf_name, _properties.properties(), _if_not_exists, _static_columns, _properties.properties()->get_id());
 
+    bool ks_uses_tablets = false;
+    try {
+        ks_uses_tablets = db.find_keyspace(keyspace()).get_replication_strategy().uses_tablets();
+    } catch (const data_dictionary::no_such_keyspace& e) {
```

Comment:
why no_such_keyspace can inherit from invalid_request_exception instead?

Reply:
invalid_request_exception is a CQL-specific exception... What is the no_such_keyspace happens in Alternator, or inside repair, or anything not in a CQL request?

Also, how would find_keyspace() know how to format the error string? E.g., it doesn't know that it was called in CREATE TABLE so can't say that in the error message. Sounds to me that the caller needs to ammend the error string anyway (replace it, or prepend it with some prefix) so it anyway needs to catch the exception. No?

Reply:
In this patch I don't see added string mentioning create table - looks like a generic one.

But ok, maybe proposed hierarchy wouldn't be correct

Reply:
> In this patch I don't see added string mentioning create table - looks like a generic one.

ׁHe chose the sentence "Unknown keyspace XYZ", but could have wanted to use different wording, e.g., "Couldn't create table in non-existent keyspace XYZ" or some other wording. I think you can't count on some generic "unknown keyspace xyz" wording to be suitable stand-alone in all cases.

Reply:
> In this patch I don't see added string mentioning create table - looks like a generic one.

> ׁHe chose the sentence "Unknown keyspace XYZ", but could have wanted to use different wording, e.g., "Couldn't create table in non-existent keyspace XYZ"

To me this message looked like it's enough, given the context in which it executes, consider:
```
cqlsh> CREATE TABLE not_existing_ks.my_table (id VARCHAR PRIMARY KEY, field_a VARCHAR, field_b VARCHAR);
InvalidRequest: Error from server: code=2200 [Invalid query] message="Unknown keyspace not_existing_ks"
```

> why no_such_keyspace can't inherit from invalid_request_exception instead

I think @nyh gave a solid [answer](https://github.com/scylladb/scylladb/pull/20404/commits/2dbb94aa79eeb7214551be113c28786b8a762ff7#r1745079164) , but it remains "tribal knowledge". I wanted to warn about it by adding an extra comment next to both exception classes.

Reply:
@ptrsmrn I don't have a strong feeling if this message is enough or not, I just pointed out that other callers might want to use a different message, and in general they need to be able to.
Whether this message is "enough" please note that the error message on its own, **without** the request, looks like:
```
InvalidRequest: Error from server: code=2200 [Invalid query] message="Unknown keyspace not_existing_ks"
```
doesn't say which operation was attempted. I think it's fine because the caller would know which operation it attempted, but perhaps in a large application if just these error messages are logged, it will be hard to guess which operation was attempted on this non-existant not_existing_ks.

In any case, I think this error message is fine, and did not ask to change it. I just wanted that if someone in the future will want to change it, I want to be able to and didn't want some keyspace-lookup function to decide the full user-visible error message (and in any case, as I noted this whole discussion is moot because in Alternator, the exception type needs to be different anyway).

Reply:
OK, why not to make our lives easier and provide the context in the log msg - I changed the msg to: "Cannot create a table in a non-existent keyspace: XYZ"